src/reference/02-DetailTopics/05-Plugins-and-Best-Practices/03-Plugins-Best-Practices.md

```diff
@@ -270,6 +294,68 @@ lazy val app = (project in file("app"))
   .settings(inConfig(Test)(ObfuscatePlugin.baseObfuscateSettings))
 ```
 
+### Scoping advices
+
+In general, if a plugin provides keys (settings and tasks) with the widest scoping,
+and refer to them with the narrowest scoping, it will give the maximum flexibility to the build users.
+
+#### Provide default values in `globalSettings`
+
+If your settings or task do not transitively depend on a project-level settings
+(such as `baseDirectory`, `compile`, etc), provide default values in `globalSettings`.
+
+For example, in sbt's Defaults keys related to publishing such as `licenses`, `developers`,
+and `scmInfo` are all defined at the `Global` scope, typically to empty values like `Nil` and `None`.
+
+```scala
+package sbtobfuscate
+
+import sbt._, Keys._
+
+object ObfuscatePlugin extends sbt.AutoPlugin {
+  override def requires = plugins.JvmPlugin
+  override def trigger = allRequirements
+
+  object autoImport {
+    lazy val obfuscate = taskKey[Seq[File]]("obfuscate the source")
+    lazy val obfuscateOption = settingKey[ObfuscateOption]("options to configure obfuscate")
+  }
+  import autoImport._
+  override lazy val globalSettings = Seq(
+    obfuscateOption := ObfuscateOption()
+  )
+
+  override lazy val projectSettings = inConfig(Compile)(
+    obfuscate := {
+      Obfuscate(
+        (obfuscate / sources).value,
+        (obfuscate / obfuscateOption).value
+      )
+    },
+    obfuscate / sources := sources.value
+  )
+}
+
+// core feature implemented here
+object Obfuscate {
+  def apply(sources: Seq[File], opt: ObfuscateOption): Seq[File] = {
+    sources
+  }
+}
+```
+
+In the above, `obfuscateOption` is set a default made-up value in the `globalSettings`;
+but is used as `(obfuscate / obfuscateOption)` in the `projectSettings`.
+This lets the user either set `obfuscate / obfuscateOption` at a particular subproject level,
+or scoped to `ThisBuild` affecting all subprojects:
+
+```scala
+ThisBuild / obfuscate / obfuscateOption := ObfuscateOption().withX(true)
+```
+
+This requires in-depth knowledge to know that RHS are all contained in
+global scope (pure values and other global settings), but it's a useful technique.
```

Comment:
What do you mean here? It doesn't make sense to me.

Reply:
I am saying that since there's no dynamic dispatch, you the plugin author need to know/guess that all the stuff in RHS are actually safe to call in global scope. For example `compile` is probably not good but what about `libraryDependencies` etc.

Reply:
Ah, I see.